src/ripple/server/impl/BaseWSPeer.h

```diff
@@ -0,0 +1,303 @@
+
+    Impl&
+    impl()
+    {
+        return *static_cast<Impl*>(this);
+    }
+
+    void
+    on_write_sb(error_code const& ec);
+
+    void
+    do_write();
+
+    void
+    on_write(error_code const& ec);
+
+    void
+    on_write_fin(error_code const& ec);
+
+    void
+    do_read();
+
+    void
+    on_read(error_code const& ec);
+
+    void
+    on_close(error_code const& ec);
+
+    virtual
+    void
+    do_close() = 0;
+};
+
+//------------------------------------------------------------------------------
+
+template<class Impl>
+template<class Body, class Allocator>
+BaseWSPeer<Impl>::BaseWSPeer(
+    Port const& port,
+    Handler& handler,
+    endpoint_type remote_address,
+    beast::http::parsed_request<Body, Allocator>&& request,
+    boost::asio::io_service& io_service,
+    beast::Journal journal)
+    : BasePeer<Impl>(port, handler, remote_address,
+        io_service, journal)
+    , request_(std::move(request))
+{
+    impl().ws_.set_option(
+        beast::wsproto::frag_size(256));
+}
+
+template<class Impl>
+void
+BaseWSPeer<Impl>::run()
+{
+    if(! strand_.running_in_this_thread())
+        return strand_.post(std::bind(
+            &BaseWSPeer::run, impl().shared_from_this()));
+    impl().ws_.set_option(beast::wsproto::decorate(identity{}));
+    using namespace beast::asio;
+    impl().ws_.async_accept(request_, strand_.wrap(std::bind(
+        &BaseWSPeer::on_write_sb, impl().shared_from_this(),
+            placeholders::error)));
+}
+
+template<class Impl>
+void
+BaseWSPeer<Impl>::send(std::shared_ptr<WSMsg> w)
+{
+    if(! strand_.running_in_this_thread())
+        return strand_.post(std::bind(
+            &BaseWSPeer::send, impl().shared_from_this(),
+                std::move(w)));
+    wq_.emplace_back(std::move(w));
+    if(wq_.size() == 1)
+        on_write({});
+}
+
+template<class Impl>
+void
+BaseWSPeer<Impl>::close()
+{
+    if(! strand_.running_in_this_thread())
+        return strand_.post(std::bind(
+            &BaseWSPeer::close, impl().shared_from_this()));
+    if(wq_.size() > 0)
+        do_close_ = true;
+    else
+        impl().ws_.async_close({}, strand_.wrap(std::bind(
+            &BaseWSPeer::on_close, impl().shared_from_this(),
+                beast::asio::placeholders::error)));
+}
+
+template<class Impl>
+void
+BaseWSPeer<Impl>::on_write_sb(error_code const& ec)
+{
+    if(ec)
+        return fail(ec, "write_resp");
+    do_read();
+}
+
+template<class Impl>
+void
+BaseWSPeer<Impl>::do_write()
+{
+    if(! strand_.running_in_this_thread())
+        return strand_.post(std::bind(
+            &BaseWSPeer::do_write, impl().shared_from_this()));
+    on_write({});
+}
+
+template<class Impl>
+void
+BaseWSPeer<Impl>::on_write(error_code const& ec)
+{
+    if(ec)
+        return fail(ec, "write");
+    auto& w = *wq_.front();
+    using namespace beast::asio;
+    auto const result = w.prepare(65536,
+        std::bind(&BaseWSPeer::do_write,
+            impl().shared_from_this()));
```

Comment:
I can't figure out how the `resume` function is used in `prepare`. I can't find an overload of `prepare` that makes use of it, but here we're passing a value. Can you point me to an implementation of `prepare` that uses `resume` so I can understand what's going on?